### ISSUE TITLE
False-positive `SMART_CAST_NEEDED` with mutable properties

### ISSUE THREAD

**Issue by petertrr** [analysis-dev/diktat#1168]
## Describe the bug
If a property is mutable an is used in `when`, then smart cast is impossible - it leads to a compilation error:
```kotlin
fun SomeClass.foo() = when (mutableProperty) {
    is Foo -> (mutableProperty as Foo)  // smart cast is required 'because 'mutableProperty' is a property that has open or custom getter'
}
```

## Expected behavior
Diktat should not trigger on such cases

## Observed behavior
Explicit casting is removed, leading to a compilation error.

## Environment information
* diktat version: 1.0.1
